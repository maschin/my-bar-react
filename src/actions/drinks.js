import moment from 'moment/moment';
import database from '../firebase/firebase';

//ADD DRINK
export const addDrink = (drink) => ({
    type: 'ADD_DRINK',
    drink
});

export const startAddDrink = (drinkData = {}) => {
    return (dispatch, getState) => {
        const {
            title = '',
            category = 'Undefined',
            vol = '',
            alc = '',
            price = '',
            aged = '',
            description = '',
            isEmpty = false,
            isOpen = false,
            boughtAt = 0,
            bottledAt = 0,
            bestBefore = 0,
            createdAt = 0,
            updatedAt = 0
        } = drinkData;

        const uid = getState().auth.uid;
        const drink = {
            title, category, vol, alc, price, aged, description, isEmpty, isOpen, boughtAt, bottledAt, bestBefore, createdAt, updatedAt
        };

        return database.ref(`users/${uid}/drinks`).push(drink).then((ref) => {
            dispatch(addDrink({
                id: ref.key,
                ...drink
            }));
        });
    }
};

//EDIT DRINK
export const editDrink = (id, updates) => ({
    type: 'EDIT_DRINK',
    id,
    updates
});

export const startEditDrink = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;

        return database.ref(`users/${uid}/drinks/${id}`).update(updates).then(() => {
            dispatch(editDrink(id, updates));
        });
    }
};

//OPEN DRINK
export const openDrink = ({ id } = {}) => {
    return (dispatch, getState) => {
        const updates = { isOpen: true, openedAt: moment().valueOf() };
        return dispatch(startEditDrink(id, updates));
    }
};

//EMPTY DRINK
export const emptyDrink = ( { id } = {}) => {
    return (dispatch, getState) => {
        const updates = { isEmpty: true, emptiedAt: moment().valueOf() };
        return dispatch(startEditDrink(id, updates));
    }
};

//REMOVE DRINK
export const removeDrink = ( { id } = {}) => ({
    type: 'REMOVE_DRINK',
    id
});

export const startRemoveDrink = ( { id } = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;

        return database.ref(`users/${uid}/drinks/${id}`).remove().then(() => {
            dispatch(removeDrink({ id }))
        });
    }
};

//SET DRINKS
export const setDrinks = (drinks) => ({
    type: 'SET_DRINKS',
    drinks
});

export const startSetDrinks = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;

        return database.ref(`users/${uid}/drinks`).once('value').then((snapshot) => {
           const drinks = [];

           snapshot.forEach((childSnapshot) => {
               drinks.push({
                  id: childSnapshot.key,
                   ...childSnapshot.val()
               });
           });

           dispatch(setDrinks(drinks));
        });

    }
}