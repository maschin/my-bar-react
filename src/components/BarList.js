import React from 'react';
import { connect } from 'react-redux';

import DrinkListItem from '../components/DrinkListItem';

export const BarList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-desktop">Drink</div>
            <div className="show-for-desktop">Data</div>
        </div>
        <div className="list-body">
            {
                props.drinks.length === 0 ? (
                    <div className="list-item list-item--message">
                        <span>Your bar is totally empty!</span>
                    </div>
                ) : (
                    props.drinks.map((drink) => (
                       <DrinkListItem key={drink.id} {...drink} />
                    ))
                )
            }
        </div>
    </div>
);

const mapStateToProps = (state) => {
    return {
        drinks: state.drinks
    };
};

export default connect(mapStateToProps)(BarList);