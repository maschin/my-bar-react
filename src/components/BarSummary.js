import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

export const BarSummary = ({ bottlesCount }) => {
    const bottlesWord = bottlesCount === 1 ? 'bottle' : 'bottles';

    return(
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">
                    You have total <span>{bottlesCount}</span> {bottlesWord} in your bar.
                </h1>
                <div className="page-header__actions">
                    <Link className="button" to="/create">Add Drink</Link>
                </div>
            </div>
        </div>
    );
};

const mapStateToProps = (state) => {
    return {
        drinks: state.drinks,
        bottlesCount: state.drinks.length
    };
};

export default connect(mapStateToProps)(BarSummary);