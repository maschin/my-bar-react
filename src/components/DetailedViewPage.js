import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import {
    openDrink,
    emptyDrink
} from '../actions/drinks';

export class DetailedView extends React.Component {
    onOpen = () => {
        this.props.openDrink(this.props.drink);
    }
    onMakeEmpty = () => {
        this.props.emptyDrink(this.props.drink);
    }
    render(){
        return(
            <div>
                {
                    this.props.drink ? (
                        <div>
                            <div className="page-header">
                                <div className="content-container">
                                    <div className="page-header__actions">
                                        <Link className="button" to={`/edit/${this.props.drink.id}`}>Edit</Link>
                                        { !this.props.drink.isOpen ? (
                                            <button id="openDrink" className="button" onClick={this.onOpen}>Open</button>
                                        ) : (

                                            !this.props.drink.isEmpty && <button id="emptyDrink" className="button" onClick={this.onMakeEmpty}>Empty</button>

                                        )}
                                    </div>
                                </div>
                            </div>
                            <div className="content-container">
                                <h1>{ this.props.drink.title }</h1>
                                <ul>
                                    <li>{ this.props.drink.category }</li>
                                    <li>{ this.props.drink.vol } L</li>
                                    <li>{ this.props.drink.alc } %</li>
                                    <li>{ this.props.drink.price }</li>
                                </ul>
                            </div>
                        </div>
                    ) : (
                        <h1 className="page-header__title">Record with <span>id:{this.props.match.params.id}</span> not found!</h1>
                )}
            </div>
        )
    }
}

const mapStateToProps = (state, props) => {
    return {
        drink: state.drinks.find((drink) => (drink.id === props.match.params.id))
    }
}

const mapDispatchToProps = (dispatch) => ({
        openDrink: (drink) => dispatch(openDrink(drink)),
        emptyDrink: (drink) => dispatch(emptyDrink(drink))
})

export default connect(mapStateToProps, mapDispatchToProps)(DetailedView);