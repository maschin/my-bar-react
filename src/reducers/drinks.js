const drinksReducerDefaultState = [
    {
        id: 'sdfsdf',
        title: 'My bottle',
        category: 'Beer',
        vol: 0.5,
        alc: 5.6,
        price: 450,
        aged: 12,
        boughtAt: 123123,
        bottledAt: 12312312,
        bestBefore: 12312312,

    },
    {
        id: 'sdfsdf',
        title: 'My bottle',
        category: 'Beer',
        vol: 0.5,
        alc: 5.6,
        price: 450,
        aged: 12,
        boughtAt: 123123,
        bottledAt: 12312312,
        bestBefore: 12312312,

    }
];

export default (state = drinksReducerDefaultState, action) => {
    switch (action.type){
        case 'SET_DRINKS':
            return action.drinks;
        default:
    }       return state;
}