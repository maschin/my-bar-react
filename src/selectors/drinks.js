//Get visible drinks

export default (drinks, { text, sortBy }) => {
    return drinks.filter((drink) => {
        const textMatch = drink.title.trim().toLowerCase().includes(text.trim().toLowerCase()) || drink.category.trim().toLowerCase().includes(text.trim().toLowerCase());

        return textMatch;
    }).sort((a, b) => {
        switch(sortBy){
            default:
                return a.createdAt < b.createdAt ? 1 : -1;
        }
    })
}
