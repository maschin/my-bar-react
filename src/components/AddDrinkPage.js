import React from 'react';
import { connect } from 'react-redux';

import DrinkForm from '../components/DrinkForm';
import { startAddDrink } from '../actions/drinks';

export class AddDrinkPage extends React.Component {
    onSubmit = (drink) => {
        this.props.startAddDrink(drink);
        this.props.history.push('/');
    };
    render(){
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add Bottle</h1>
                    </div>
                </div>
                <div className="content-container">
                    <DrinkForm
                        actionTitle='Add Bottle'
                        onSubmit={this.onSubmit}
                    />
                </div>
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => ({
    startAddDrink: (drink) => dispatch(startAddDrink(drink))
});

export default connect(undefined, mapDispatchToProps)(AddDrinkPage);