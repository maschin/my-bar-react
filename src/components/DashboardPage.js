import React from 'react';
import BarList from '../components/BarList';
import BarSummary from '../components/BarSummary';

export const DashboardPage = () => (
    <div>
        <BarSummary />
        <BarList />
    </div>
);

export default DashboardPage;