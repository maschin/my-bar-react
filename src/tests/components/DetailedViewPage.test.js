import React from 'react';
import { shallow } from 'enzyme';
import { DetailedView } from '../../components/DetailedViewPage';
import drinks from '../fixtures/drinks';

test('should render detailed view correctly with open and without empty button', () => {
   const wrapper = shallow(<DetailedView drink={drinks[1]} />);
   expect(wrapper).toMatchSnapshot();
});

test('should render detailed view correctly with empty and without open button', () => {
    const wrapper = shallow(<DetailedView drink={drinks[2]} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render detailed view correctly without empty & open buttons', () => {
    const wrapper = shallow(<DetailedView drink={drinks[3]} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render detailed view with incorrect id correctly', () => {
    const params = { id: 123 };
    const match = { params };
    const wrapper = shallow(<DetailedView match={match}/>)
    expect(wrapper).toMatchSnapshot();
});

test('should handle on edit', () => {
    const wrapper = shallow(<DetailedView drink={drinks[1]} />);
    expect(wrapper.find('Link').prop('to')).toBe(`/edit/${drinks[1].id}`);
});

test('should handle on Open', () => {
    const openDrink = jest.fn();
    const wrapper = shallow(<DetailedView drink={drinks[1]} openDrink={openDrink} />);
    wrapper.find('#openDrink').simulate('click');
    expect(openDrink).toHaveBeenCalledWith(drinks[1]);
});

test('should handle on Empty', () => {
    const emptyDrink = jest.fn();
    const wrapper = shallow(<DetailedView drink={drinks[2]} emptyDrink={emptyDrink} />);
    wrapper.find('#emptyDrink').simulate('click');
    expect(emptyDrink).toHaveBeenCalledWith(drinks[2]);
});