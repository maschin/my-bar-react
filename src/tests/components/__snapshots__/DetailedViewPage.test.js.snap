// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render detailed view correctly with empty and without open button 1`] = `
<div>
  <div>
    <div
      className="page-header"
    >
      <div
        className="content-container"
      >
        <div
          className="page-header__actions"
        >
          <Link
            className="button"
            replace={false}
            to="/edit/leffe"
          >
            Edit
          </Link>
          <button
            className="button"
            id="emptyDrink"
            onClick={[Function]}
          >
            Empty
          </button>
        </div>
      </div>
    </div>
    <div
      className="content-container"
    >
      <h1>
        Leffe Ruby
      </h1>
      <ul>
        <li>
          Beer
        </li>
        <li>
          0.25
           L
        </li>
        <li>
          5
           %
        </li>
        <li>
          11.12
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`should render detailed view correctly with open and without empty button 1`] = `
<div>
  <div>
    <div
      className="page-header"
    >
      <div
        className="content-container"
      >
        <div
          className="page-header__actions"
        >
          <Link
            className="button"
            replace={false}
            to="/edit/laphroaigh"
          >
            Edit
          </Link>
          <button
            className="button"
            id="openDrink"
            onClick={[Function]}
          >
            Open
          </button>
        </div>
      </div>
    </div>
    <div
      className="content-container"
    >
      <h1>
        Laphroaig QA Cask Double Matured
      </h1>
      <ul>
        <li>
          Whisky
        </li>
        <li>
          1
           L
        </li>
        <li>
          40
           %
        </li>
        <li>
          4804
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`should render detailed view correctly without empty & open buttons 1`] = `
<div>
  <div>
    <div
      className="page-header"
    >
      <div
        className="content-container"
      >
        <div
          className="page-header__actions"
        >
          <Link
            className="button"
            replace={false}
            to="/edit/whisky"
          >
            Edit
          </Link>
        </div>
      </div>
    </div>
    <div
      className="content-container"
    >
      <h1>
        My bottle II
      </h1>
      <ul>
        <li>
          Whisky
        </li>
        <li>
          0.7
           L
        </li>
        <li>
          42
           %
        </li>
        <li>
          12
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`should render detailed view with incorrect id correctly 1`] = `
<div>
  <h1
    className="page-header__title"
  >
    Record with 
    <span>
      id:
      123
    </span>
     not found!
  </h1>
</div>
`;
